clippy_lints/src/modulo_arithmetic.rs

```diff
@@ -0,0 +1,102 @@
+use crate::consts::{constant, Constant};
+use crate::utils::{sext, span_help_and_lint};
+use if_chain::if_chain;
+use rustc::declare_lint_pass;
+use rustc::hir::*;
+use rustc::lint::{LateContext, LateLintPass, LintArray, LintPass};
+use rustc::ty::{self};
+use rustc_session::declare_tool_lint;
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for modulo arithemtic.
+    ///
+    /// **Why is this bad?** The results of moudlo (%) operation migth differ
+    /// depending on the language, when negative numbers are involved.
+    /// If you interop with different languages it might be beneficial
+    /// to double check all places that use modulo arithmetic.
+    ///
+    /// For example, in Rust `17 % -3 = 2`, but in Python `17 % -3 = -1`.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    /// ```rust
+    /// let x = -17 % 3;
+    /// ```
+    pub MODULO_ARITHMETIC,
+    nursery,
+    "any modulo arithmetic statement"
+}
+
+declare_lint_pass!(ModuloArithmetic => [MODULO_ARITHMETIC]);
+
+fn is_negative<'a, 'tcx>(c: &ptr::P<rustc::hir::Expr>, cx: &LateContext<'a, 'tcx>, expr: &'tcx Expr) -> Option<bool> {
+    match constant(cx, cx.tables, c) {
+        Some((Constant::Int(v), _)) => match cx.tables.expr_ty(expr).kind {
+            ty::Int(ity) => {
+                return Some(sext(cx.tcx, v, ity) < 0);
+            },
+            ty::Uint(_) => {
+                return Some(false);
+            },
+            _ => {},
+        },
+        Some((Constant::F32(f), _)) => {
+            return Some(f < 0.0);
+        },
+        Some((Constant::F64(f), _)) => {
+            return Some(f < 0.0);
+        },
+        _ => {},
+    }
+    None
+}
+
+fn might_have_negative_value(t: &ty::TyS<'_>) -> bool {
+    (t.is_integral() && t.is_signed()) || t.is_floating_point()
+}
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for ModuloArithmetic {
+    fn check_expr(&mut self, cx: &LateContext<'a, 'tcx>, expr: &'tcx Expr) {
+        match &expr.kind {
+            ExprKind::Binary(op, lhs, rhs) | ExprKind::AssignOp(op, lhs, rhs) => {
+                if let BinOpKind::Rem = op.node {
+                    let lhs_negative = is_negative(lhs, cx, expr);
+                    let rhs_negative = is_negative(rhs, cx, expr);
+                    if lhs_negative.is_some() && rhs_negative.is_some() {
+                        // Operands are consts
+                        if_chain! {
+                            if let Some(lhs_negative) = lhs_negative;
+                            if let Some(rhs_negative) = rhs_negative;
+                            if lhs_negative == !rhs_negative;
```

Comment:
```suggestion
                            if lhs_negative != rhs_negative;
```

We don't have a lint for this? 😮

Reply:
Wait, these are bools. So we could use a XOR here
```suggestion
                            if lhs_negative ^ rhs_negative;
```

Reply:
Fixed in https://github.com/rust-lang/rust-clippy/pull/4867/commits/bc318e9185e5a0404d06cd31d7ab7c0750c922ec

Reply:
Actually, I might start working on a lint that will suggest using xor in similar situations...